prov/efa/test/efa_unit_test_hmem.c

```diff
@@ -0,0 +1,48 @@
+#include "efa_unit_tests.h"
+
+
+#if HAVE_NEURON
+/**
+ * @brief Verify when neuron_alloc failed (return null),
+ * efa_domain_open, which call efa_hmem_support_status_update_neuron
+ * when HAVE_NEURON=1, will still return 0 but leave 
+ * hmem_support_status[FI_HMEM_NEURON].initialized and 
+ * hmem_support_status[FI_HMEM_NEURON].p2p_supported as false.
+ */
+void test_efa_hmem_support_status_update_neuron()
+{
+        int ret;
+        struct efa_resource resource = {0};
+        struct efa_domain *efa_domain;
+
+        resource.hints = efa_unit_test_alloc_hints(FI_EP_RDM);
+        assert_non_null(resource.hints);
+
+        ret = fi_getinfo(FI_VERSION(1, 14), NULL, NULL, 0ULL, resource.hints, &resource.info);
+        assert_int_equal(ret, 0);
+
+        ret = fi_fabric(resource.info->fabric_attr, &resource.fabric, NULL);
+        assert_int_equal(ret, 0);
+
+        hmem_ops[FI_HMEM_NEURON].initialized = true;
+        g_device_list[0].device_caps |= EFADV_DEVICE_ATTR_CAPS_RDMA_READ;
+        g_efa_unit_test_mocks.neuron_alloc = &efa_mock_neuron_alloc_return_null;
+
+        ret = fi_domain(resource.fabric, resource.info, &resource.domain, NULL);
+        assert_int_equal(ret, 0);
+
+        efa_domain = container_of(resource.domain, struct efa_domain,
+				  util_domain.domain_fid.fid);
+        assert_false(efa_domain->hmem_support_status[FI_HMEM_NEURON].initialized);
+        assert_false(efa_domain->hmem_support_status[FI_HMEM_NEURON].p2p_supported);
+        
+        /* need to reset this to avoid dlclose (in ofi_hmem_cleanup) being called */
+        hmem_ops[FI_HMEM_NEURON].initialized = false;
```

Comment:
Do you need to also reset `g_device_list[0].device_caps`?

Reply:
Yes, added in updated revision. Thank you!

Reply:
I think you need to save original value of `g_device_list[0].device_caps`?

Reply:
I see, save and recover the original value in updated revision.

Reply:
The problem with resetting value after `assert_xxx` is that it won't be called if assertion fails. We have 2 options:
1. Still reset state inside the test, but reset the state immediately after it has taken effect and before assertions
2. Reset the state outside the test, in a cleanup hook, e.g. `efa_unit_test_mocks_reset` or similar.

Reply:
For number 2, how do you know the original value before the test modifying the variables?

Reply:
I will do number 1 for this PR.

Reply:
For option 2, I think the easier(cleaner) way is to make sure every test starts with the same global state, including `g_device_list`. Basically we want to take out the guess work and hard code a value.